### PR TITLE
Add Sort Control 

### PR DESCRIPTION
**Pull request by irvingoujAtDevolution**
Oid Specification: https://www.alvestrand.no/objectid/1.2.840.113556.1.4.473.html

Wireshark Recognize it, this changed is based on,https://github.com/kanidm/ldap3/pull/42, which should be merged first.

image

image

### CODE REVIEW

File: proto/Cargo.toml
Comment (CBenoit) [inline proto/Cargo.toml:29]: Question: typically I would avoid bringing a new proc macro crate / utility library unless it’s really worth it. In this case, bringing in all of this just for one enum doesn’t sounds very compelling. That being said, the heaviest dependency, syn, is already part of the dependency graph (via thiserror and serde), so it’s not too bad. I’m not asking you to remove it if you prefer using it and if @Firstyear is good with this additional dependency.

File: proto/src/control.rs
Comment (CBenoit) [inline proto/src/control.rs:509]: Nitpick: we typically don’t call `clone` on `Copy` types (unless writing generic code)

Suggestion:
```diff
@@ -496,14 +496,15 @@ impl TryFrom<StructureTag> for LdapControl {
                     .and_then(|t| t.expect_primitive())
                     .ok_or(LdapProtoError::ControlBer)?;

-                let code = value.first().ok_or(LdapProtoError::ControlBer)?;
+                let result_code = value
+                    .first()
+                    .copied()
+                    .and_then(|code| ServerSortResultCode::try_from(code).ok())
+                    .ok_or(LdapProtoError::ControlBer)?;

                 Ok(LdapControl::ServerSortResult {
                     sort_result: ServerSortResult {
-                        result_code: code
-                            .clone()
-                            .try_into()
-                            .map_err(|_| LdapProtoError::ControlBer)?,
+                        result_code,
                         attribute_type: None, // TODO!
                     },
                 })
```

File: client/src/addirsync.rs
Comment (Firstyear) [inline client/src/addirsync.rs:4]: This depends on the control move PR?

File: proto/Cargo.toml
Comment (Firstyear) [reply proto/Cargo.toml:29]: My attitude is "bring in as few deps as possible, unless they really help and solve a problem that otherwise would be painful. 

I think in this case, there may be other places where num_enum could be used in the project, so it may be worth and as you say, it's already got most of it's deps in the code base. So I'm okay with it here.

File: proto/src/control.rs
Comment (Firstyear) [reply proto/src/control.rs:509]: clippy complains if you do .clone() on a copy type too.

File: client/src/addirsync.rs
Comment (CBenoit) [reply client/src/addirsync.rs:4]: Looks like it does.

> this change is based on https://github.com/kanidm/ldap3/pull/42, which should be merged first.

@irvingoujAtDevolution I suggest to mark the PR as draft to avoid confusion. Ideally, the diff should be generated against the base, but in this case I believe it would not be possible because the base is in our fork.

**Comment by irvingoujAtDevolution** [comment]
Thanks, will update this one once #42 is merged.

File: proto/Cargo.toml
Comment (irvingoujAtDevolution) [reply proto/Cargo.toml]: updated

File: proto/src/control.rs
Comment (irvingoujAtDevolution) [reply proto/src/control.rs]: updated

File: proto/src/control.rs
Comment (CBenoit) [inline proto/src/control.rs:94]: note: Typically, `TryFrom<u8>` is preferred over `TryFrom<&u8>` (`u8` being a `Copy` type)

File: proto/src/control.rs
Comment (irvingoujAtDevolution) [reply proto/src/control.rs]: updated, you are right.

**Review by Firstyear — APPROVED** [review]

**Comment by Firstyear** [comment]
Great work as always @irvingoujAtDevolution, I'll leave it to @CBenoit to give their review too.

**Review by CBenoit — APPROVED** [review]
No concern for me either :+1:
Thank you! @irvingoujAtDevolution @Firstyear

**Pull request merged**
